### PR TITLE
Support better configurability for pivotal flow [#169620136] [Fixes #23, #20]

### LINKED ISSUE #23: Better configurability
* Possibly use something like [cosmiconfig](https://github.com/davidtheclark/cosmiconfig) to be able to set up project IDs etc
* Allow picking stories from specified project ID

### PR DESCRIPTION
**Pull request by devoficial**

### CODE REVIEW

File: src/commands/init/utils.ts [hunk 1]
```diff
@@ -20,15 +26,55 @@ To get started with pivotal-flow, run the following commands in your current ter
 {dim.italic You can also add them to your profile (~/.bash_profile, ~/.zshrc, ~/.profile, or ~/.bashrc) so the environment variables are automatically added for all new terminal sessions.}
 `);
 
+export interface PivotalFlowConfig {
+  projectName: string;
+  projectId: number;
+  default?: boolean;
```
Comment (rheaditi) [inline src/commands/init/utils.ts]: Isn't default a keyword in ES? 🙈

Comment (devoficial) [reply src/commands/init/utils.ts]: changed it to ```isDefault```

**Comment by rheaditi** [comment]
@devoficial Good job on the PR! 😁👍 

Few suggestions:
* Can we use [cosmiconfig](https://github.com/davidtheclark/cosmiconfig) to manage the configuration. It allows us to use an API rather than needing to look for config in multiple places.. ? It'll also allow using any format for the file.
* Once we start using the details from the config file we can remove the extra instructions, namely this part:
image

File: src/utils/pivotal/client.ts [hunk 2]
```diff
@@ -135,17 +141,32 @@ export default class PivotalClient {
...
       },
       { progress: 'Fetching stories' }
     );
   }
 
+  /**
+   * Fetch a project details for a specified projectId
+   * @param {string} projectId - pivotal projectId
+   */
+  async getProjectDetails(projectId: string) {
```
Comment (rheaditi) [inline src/utils/pivotal/client.ts]: [minor, naming, optional]
getProject~Details~()

File: src/utils/pivotal/types.ts [hunk 3]
```diff
@@ -143,6 +143,11 @@ export interface GetStoriesResponse {
   query: string;
 }
 
+export interface GetProjectDetailsResponse {
```
Comment (rheaditi) [inline src/utils/pivotal/types.ts]: [minor, naming]: Can this just be `PivotalProject`?

Comment (rheaditi) [reply src/utils/pivotal/types.ts]: [minor, optional] can we add some more of the fields from https://www.pivotaltracker.com/help/api/rest/v5#project_resource ?

Comment (devoficial) [reply src/utils/pivotal/types.ts]: what are the fields do we require, I don't know which all are we need ??

Comment (devoficial) [reply src/utils/pivotal/types.ts]: already there is a ```PivotalProjectResponse``` will use that interface

**Comment by devoficial** [comment]
> @devoficial Good job on the PR! 😁👍
> 
> Few suggestions:
> 
> * Can we use [cosmiconfig](https://github.com/davidtheclark/cosmiconfig) to manage the configuration. It allows us to use an API rather than needing to look for config in multiple places.. ? It'll also allow using any format for the file.
> * Once we start using the details from the config file we can remove the extra instructions, namely this part:
>   image

File: src/utils/fs.ts [hunk 4]
```diff
@@ -2,3 +2,7 @@ import { promisify } from 'util';
 import fs from 'fs';
 
 export const readFile = promisify(fs.readFile);
+
+export enum configFileName {
+  PIVOTAL_CONFIG_FILE = '.pivotal-flow-config.json',
```
Comment (rheaditi) [inline src/utils/fs.ts]: Let's move this to a constants file?

File: src/commands/init/constants.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export enum configFileName {
```
Comment (rheaditi) [inline src/commands/init/constants.ts]:
```suggestion
export enum CONFIG {
```

File: src/commands/init/utils.ts [hunk 6]
```diff
@@ -1,45 +1,119 @@
...
+ * creates a config file at user's home directory to store pivotal projects and API token
+ * @param projectId
+ * @param apiToken
  */
-const displaySetupInstructions = ({ pivotalToken, pivotalProjectId }: SetupAnswers) =>
+export const createPivotalFlowConfigFile = async (projectId: string, apiToken: string): Promise<void> => {
+  const client = new PivotalClient({ debug: true, projectId, apiToken });
+  const projectDetails = await client.getProject();
+  const pivotalFlowConfig = createPivotalFlowConfig(projectDetails, apiToken);
+  const jsonifyConfig = JSON.stringify([pivotalFlowConfig]);
```
Comment (rheaditi) [inline src/commands/init/utils.ts]:
```suggestion
  const jsonifyConfig = JSON.stringify(pivotalFlowConfig, null, 2);
```

File: src/commands/init/utils.ts [hunk 7]
```diff
@@ -1,45 +1,119 @@
...
+
+  writeFileSync(resolve(homedir(), configFileName.PIVOTAL_CONFIG_FILE), jsonifyConfig, { encoding: 'utf8' });
```
Comment (rheaditi) [inline src/commands/init/utils.ts]: todo:
Confirm before over-writing..

File: src/commands/start/questions.ts [hunk 8]
```diff
@@ -25,6 +30,24 @@ export const PickStoryWorkflowQuestions: QuestionCollection<PickStoryWorkflowAns
   },
 ];
 
+export const PickProjectWorkflowQuestions = (
+  projects: PivotalProjectConfig[]
+): QuestionCollection<PickProjectWorkflowAnswers> => {
+  const projectChoices = projects.map((project: PivotalProjectConfig) => {
+    const { projectName: name, projectId: value } = project;
+    return { name, value };
```
Comment (rheaditi) [inline src/commands/start/questions.ts]:
```suggestion
    return { name: `${name} [${value}]`, value };
```

File: src/commands/start/utils.ts [hunk 9]
```diff
@@ -147,6 +149,25 @@ export const getWorkflow = async ({ newStory }: { newStory: boolean }): Promise<
   return selection;
 };
 
+export const selectProjectToCreateStory = async (): Promise<{ apiToken: string; projectId: string }> => {
+  const pivotalConfig = await getPivotalFlowConfig();
+  let projectConfig = { projectId: '', apiToken: '' };
+
+  if (pivotalConfig && Array.isArray(pivotalConfig)) {
```
Comment (rheaditi) [inline src/commands/start/utils.ts]: reminder: Array check :P

File: src/commands/start/utils.ts [hunk 10]
```diff
@@ -170,7 +191,7 @@ export const getStoryToWorkOn = async (
  * @see https://github.com/mokkabonna/inquirer-autocomplete-prompt#example
  */
 export const getSearchableStoryListSource = (
-  /** Current story list to search from */
+  /* Current story list to search from */
```
Comment (rheaditi) [inline src/commands/start/utils.ts]:
```suggestion
  /** Current story list to search from */
```

**Pull request merged**